### PR TITLE
[diffusion][model] Add LingBot-Video MoE 30B T2V support

### PR DESCRIPTION
**Pull request by pandalee99**
## Motivation

Add native SGLang support for **LingBot-Video MoE 30B** (`robbyant/lingbot-video-moe-30b-a3b`) — a DeepSeek-V3-style MoE text-to-video model (128 experts, 30B total / ~3B active per token, 48 MoE layers). This is the **first MoE DiT** in `multimodal_gen`, reusing SGLang's `fused_experts` Triton kernel for the expert GEMMs.

Related issue: #32336

## Modifications

- **`runtime/layers/moe.py`** (NEW): DeepSeek-V3-style MoE FFN block — `LingBotVideoRouter` (fp32, group-limited top-k, sigmoid + `e_score_correction_bias`), `LingBotVideoGroupedExperts` (w1/w2/w3 `nn.Parameter`), `LingBotVideoSparseMoeBlock` reusing `sglang.srt.layers.moe...fused_experts`. First MoE layer in the diffusion runtime. Parity-critical config: `gate_up_interleaved=False`, `inplace=False`.
- **`runtime/models/dits/lingbot_video_moe.py`** (NEW): `LingBotVideoTransformer3DModel` (joint self-attn over `[video;text]`, complex64 RoPE via `view_as_complex`, AdaLN with fp32-sensitive modules). `LingBotVideoAttention` (SDPA + complex RoPE), `LingBotVideoBlock`.
- **`configs/models/dits/lingbot_video_moe.py`** (NEW): `LingBotVideoMoEArchConfig` (30B arch + identity `param_names_mapping`).
- **`configs/pipeline_configs/lingbot_video_moe.py`** (NEW): `LingBotVideoMoEPipelineConfig` (T2V, `get_decode_scale_and_shift` reading `vae.config`).
- **`configs/sample/lingbot_video_moe.py`** (NEW): `LingBotVideoMoESamplingParams`.
- **`runtime/pipelines_core/stages/model_specific_stages/lingbot_video_moe/`** (NEW): `LingBotVideoBeforeDenoisingStage` (Hybrid monolithic: encode_prompt + Qwen3-VL + PROMPT_TEMPLATE, prepare_latents, FlowUniPC).
- **`runtime/pipelines/lingbot_video_moe.py`** (NEW): `LingBotVideoPipeline` (`EntryClass`).
- **`registry.py`** (MODIFIED): `register_configs(...)` entry.
- **`configs/models/dits/__init__.py`**, **`configs/pipeline_configs/__init__.py`**, **`configs/sample/__init__.py`** (MODIFIED): exports.

## Usage

```bash
# 32GB GPU (offload):
CUDA_VISIBLE_DEVICES=0 sglang serve robbyant/lingbot-video-moe-30b-a3b \
  --num-gpus 1 --dit-layerwise-offload --warmup-mode off

# 80GB+ GPU (resident):
CUDA_VISIBLE_DEVICES=0 sglang serve robbyant/lingbot-video-moe-30b-a3b --num-gpus 1
```

```bash
curl http://localhost:30000/v1/videos \
  -F "model=robbyant/lingbot-video-moe-30b-a3b" \
  -F "prompt=<structured JSON caption>" \
  -F "size=320x192" -F "num_frames=241" -F "fps=24" \
  -F "num_inference_steps=40" -F "guidance_scale=6.0" \
  -F "flow_shift=3.0" -F "seed=42"
```

> **Note:** The DiT is trained on structured JSON captions (from the prompt-rewriter). Raw natural-language prompts are out-of-distribution. The `prompt` field should carry a structured JSON caption (e.g., the rewriter's output from the upstream `assets/cases/t2v/example_1/prompt.json`).

## Accuracy Tests

Verified output precision against the upstream `LingBotVideoPipeline` (diffusers) with **identical parameters**: same prompt (structured JSON caption), seed (42), resolution (320×192), num_frames (241), steps (40), guidance_scale (6.0), shift (3.0).

| | Hardware | Time | Notes |
|---|---|---|---|
| **SGLang (this PR)** | RTX 5090 (32GB, `--dit-layerwise-offload`) | 115s | DiT blocks streamed from CPU |
| **SGLang (this PR)** | H100 80GB (resident) | 70s | No offload |
| **Upstream (diffusers)** | H100 80GB (resident) | 85s | Original `LingBotVideoPipeline` |

Both SGLang and the upstream produce coherent video (cat running on grass, 10s, 320×192, 241 frames) at the same settings.

### Output Videos

**SGLang (this PR, H100 80GB):**

https://github.com/user-attachments/assets/ff44ef06-b121-476f-aead-1d4d0adcc3ae

**SGLang (this PR, RTX 5090 32GB, `--dit-layerwise-offload`):**

https://github.com/user-attachments/assets/c0c66d51-e944-439d-8146-61d4908b3468

**Upstream `LingBotVideoPipeline` (diffusers, H100 80GB, reference):**

https://github.com/user-attachments/assets/fc38b1ad-517b-4868-8bea-8012fd4dff9b

All three videos are generated with identical parameters: same structured JSON caption ("a cat running on grass"), seed (42), resolution (320×192), num_frames (241), steps (40), guidance_scale (6.0), shift (3.0).

### Note on Minor Output Differences

The SGLang output and the upstream output show **slightly different camera movement and action trajectory**, while the scene content (the cat, the grass, the overall composition) is fully consistent with the prompt. The differences come from two BF16 numerical sources:

1. **MoE expert GEMM kernel**: SGLang uses `fused_experts` (srt Triton kernel with `gate_up_interleaved=False`, `inplace=False`), while the upstream uses `torch._grouped_mm` (or a per-expert loop). These are different implementations of the same grouped GEMM math, with BF16 accumulation-order differences that compound across 48 MoE layers × 40 denoise steps.

2. **Attention backend**: SGLang uses `F.scaled_dot_product_attention` with auto backend selection (flash SDP on Hopper), while the upstream uses diffusers `dispatch_attention_fn` with `_native_flash` (`sdpa_kernel(SDPBackend.FLASH_ATTENTION)`). The two paths use different SDP dispatch logic, introducing BF16 rounding differences before MoE routing, which accumulates over steps.

These are expected BF16 inference-level differences (not correctness bugs). The scene content, prompt adherence, and overall quality are equivalent.

- `ruff check`: All checks passed.
- `ruff format`: All files formatted.
- Follows the `sglang-diffusion-add-model` skill (Hybrid Monolithic pipeline, `EntryClass`, `ComposedPipelineBase`, Req batch contract).

## Future Work

To reduce review difficulty, this PR implements **T2V base-only, single-GPU**. Follow-up PRs will add:
- **Multi-GPU**: TP/SP/FSDP (swap plain `nn.Linear` → `ColumnParallelLinear`, SDPA → `USPAttention`; integrate complex64 RoPE with Ulysses).
- **Refiner**: the 1080p upscaling stage (the model includes a `refiner/` DiT with identical MoE architecture; wired via the low-noise-tail `t_thresh` mechanism).
- **Prompt rewriter**: the two-stage Qwen3.6-27B rewriter (raw prompt → structured JSON caption) as a preprocessing service.
- The srt bridge (`_ensure_moe_server_args`, `_ensure_srt_distributed`) is a pragmatic workaround; a framework-level fix (the `multimodal_gen` worker initializing the srt parallel_state when using `fused_experts`) would replace these.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.ai/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.ai/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #31158986364](https://github.com/sgl-project/sglang/actions/runs/31158986364)
Latest PR Test (Extra): :x: [Run #31158986144](https://github.com/sgl-project/sglang/actions/runs/31158986144)

### CODE REVIEW

File: python/sglang/multimodal_gen/runtime/models/dits/lingbot_video_moe.py [hunk 1]
```diff
@@ -0,0 +1,523 @@
...
+        self.norm_k = LingBotVideoRMSNorm(self.head_dim, norm_eps)
+        self.to_out = nn.Linear(hidden_size, hidden_size, bias=out_bias)
+
+    def forward(
+        self,
+        x: torch.Tensor,
+        rotary_emb: torch.Tensor,
+        attention_mask: torch.Tensor | None = None,
+    ) -> torch.Tensor:
+        _, _, _ = x.shape
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/models/dits/lingbot_video_moe.py]: could you adapt to USPAttention and native rotary embedding modules?

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/model_specific_stages/lingbot_video_moe/before_denoising.py [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+    "<|im_start|>assistant\n"
+)
+IMG_PROMPT_TEMPLATE = "<|vision_start|><|image_pad|><|vision_end|>"
+VIDEO_PROMPT_TEMPLATE = "<|vision_start|><|video_pad|><|vision_end|>"
+
+DEFAULT_NEGATIVE_PROMPT = '{"universal_negative": {"visual_quality": ["low quality", "worst quality", "blurry", "pixelated", "jpeg artifacts", "low resolution", "unstable color", "color flicker", "underexposed", "overexposed", "invisible subject", "subject hidden in darkness"], "artistic_style": ["painting", "illustration", "drawing", "cartoon", "3d render", "cgi", "sketch", "digital art"], "composition_and_content": ["text", "watermark", "signature", "logo", "subtitles", "pillarboxed", "side bars", "portrait image in landscape frame"], "temporal_and_motion_stability": ["flickering", "jittery", "motion blur", "temporal inconsistency", "warping", "morphing", "incoherent motion", "unnatural movement", "static object with sudden jump", "frame-to-frame inconsistency"], "material_and_structure": ["plastic-like glass", "unrealistic texture", "deformed bottle", "liquid freezing improperly", "distorted reflections"]}}'
+
+
+class LingBotVideoBeforeDenoisingStage(PipelineStage):
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/pipelines_core/stages/model_specific_stages/lingbot_video_moe/before_denoising.py]: we'd prefer using native split stages instead of aggregated stage

File: python/sglang/multimodal_gen/runtime/layers/moe.py [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+    )
+
+    try:
+        server_args = get_global_server_args()
+    except Exception:  # noqa: BLE001
+        server_args = None
+    if server_args is None:
+        server_args = _SGLANG_MOE_SERVER_ARGS
+        set_global_server_args_for_scheduler(server_args)
+    if not hasattr(server_args, "enable_deterministic_inference"):
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/layers/moe.py]: move to adjust logic of `ServerArgs`

**Comment by pandalee99** [comment]
Hey mickqian, all are done. Re-ran a dog video ("a dog running on grass",320×192/241f/40steps/gs6/shift3/seed42) .
- replaced the complex64 RoPE + manual SDPA with `NDRotaryEmbedding` + `_apply_rotary_emb` + `USPAttention`; q/k/v/out now use `ColumnParallelLinear`/`RowParallelLinear`.
- dropped the monolithic `BeforeDenoisingStage` for `LingBotVideoTextEncodingStage` + the standard latent/timestep/decoding stages. 

Speed :

| | Hardware | Time |
|---|---|---|
| sglang (this PR) | RTX 5090 32GB (`--dit-layerwise-offload`) | 110s |
| sglang (this PR) | H100 80GB (resident) | 70s |
| original diffusers | H100 80GB (resident) | 83s |

Videos:


https://github.com/user-attachments/assets/4e91db8a-9706-4507-bd64-e8ebf68d030f


https://github.com/user-attachments/assets/15035274-cbfc-4a62-b18e-106cc6a66e85


https://github.com/user-attachments/assets/de8b8b47-0304-4c48-bcef-6a2e67c9040a

File: python/sglang/multimodal_gen/runtime/layers/moe.py (hunk 3 above)
Comment (pandalee99) [reply python/sglang/multimodal_gen/runtime/layers/moe.py]: these 3 are really srt fused_experts internal knobs (not user flags, mm_gen doesn't read them), so bolting them on the shared ServerArgs felt out of place 
I moved them to a dedicated SrtMoeBridgeArgs  in gpu_worker, published into srt runtime_context at worker init. srt knobs no longer pollute the diffusion ServerArgs; the bridge itself is the explicit adjust surface. Verified only fused_experts reads them, so nothing else breaks.

**Comment by pandalee99** [comment]
Hi @mickqian, I've addressed the review comments and also added unit tests plus a GPU quality case, ready for review now.

**Review by mickqian — CHANGES_REQUESTED** [review]
The native MoE hot path needs another pass before merge.

1. `torch.cat((self.experts.w1, self.experts.w3), dim=1)` runs in every MoE block forward. At the current BF16 shape, it materializes `[128, 1536, 2048]` (768 MiB) per invocation. Across 48 blocks and 40 denoising steps this adds substantial HBM copies and allocator pressure. Please pack/load `w13` in the fused-kernel layout once, or cache it with correct invalidation for LoRA and other weight updates, instead of concatenating on every forward.

2. `text_lens.detach().cpu().tolist()` synchronizes the GPU on every denoising step, and the `B > 1` path dispatches attention once per sample. Please keep the text-length/RoPE preparation on-device or cache it for the request shape, then verify whether a batched/varlen attention path preserves parity. This will otherwise hurt CFG and batched serving latency.

The SRT TP-group bridge itself looks justified; no change requested there.

**Comment by pandalee99** [comment]
@mickqian You have reviewed this point very patiently. Thank you!
both fair points, Fixed:
  - w13 is one param loaded once at load time; the per-forward cat is gone, and layerwise offload handles it like any other param.
  - RoPE is built on-device in one shot, so the per-step sync and the per-sample loop are gone; B>1 attention is one batched call. Output is byte-identical to before.
  
5090, dog video:
  
| code   | RTX 5090 32GB |
|--------|--------------------------------------------|
| before | 110s                                       |
| after  | 105s                                       |

  Video, same as before:


https://github.com/user-attachments/assets/95e833e8-a35d-42cc-9047-05aa90ad4a16

**Review by mickqian — APPROVED** [review]

**Comment by pandalee99** [comment]
Hey,@mickqian . CI fails in test_update_weights_from_disk.py on FLUX.2 and Qwen-Image, a weight-name mapping mismatch in linear.py. This PR touches neither that file nor weights_updater.py, and LingBot is absent from the log. Is this known on main? Happy to rebase and re-run.

**Comment by pandalee99** [comment]
The current CI jobs are all failing after a few seconds. It seems like some label is missing, even though the checks appear to have passed.
